_drafts/firmware-code-size-optimization-1.md

```diff
@@ -0,0 +1,302 @@
+diff --git a/with-libc/samd21g18a_flash.ld b/with-libc/samd21g18a_flash.ld
+index a07cc82b..f49e73e2 100644
+--- a/with-libc/samd21g18a_flash.ld
++++ b/with-libc/samd21g18a_flash.ld
+@@ -3,7 +3,7 @@ OUTPUT_ARCH(arm)
+ 
+ MEMORY
+ {
+-  rom      (rx)  : ORIGIN = 0x00000000, LENGTH = 0x00040000
++  rom      (rx)  : ORIGIN = 0x00000000, LENGTH = 10800
+   ram      (rwx) : ORIGIN = 0x20000000, LENGTH = 0x00008000
+ }
+```
+
+We end up overflowing our flash!
+
+```terminal
+francois-mba:with-libc francois$ make
+[...]
+/usr/local/Cellar/arm-none-eabi-gcc/8-2018-q4-major/gcc/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld:
+build/with-libc.elf section `.data' will not fit in region `rom'
+/usr/local/Cellar/arm-none-eabi-gcc/8-2018-q4-major/gcc/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld:
+region `rom' overflowed by 104 bytes
+collect2: error: ld returned 1 exit status
+make: *** [build/with-libc.elf] Error 1
+```
+
+The 104 bytes overflow hints at the cause: we are overflowing our flash by
+the size of our `data` section. This is because initialization values for our
+initialized static variables are stored in flash as well.
+
+To avoid confusion, I prefer to use a small script to compute the sizes:
+
+```bash
+#!/bin/bash
+
+if [  $# -le 2 ]
+then
+    echo "This script requires 3 arguments."
+    echo -e "\nUsage:\nget-fw-size FILE MAX_FLASH_SIZE MAX_RAM_SIZE \n"
+    exit 1
+fi
+
+file=$1
+max_flash=$2
+max_ram=$3
+
+function print_region() {
+    size=$1
+    max_size=$2
+    name=$3
+
+    if [[ $max_size == 0x* ]];
+    then
+        max_size=$(echo ${max_size:2})
+        max_size=$(( 16#$max_size ))
+    fi
+
+    pct=$(( 100 * $size / $max_size ))
+    echo "$name used: $size / $max_size ($pct%)"
+}
+
+raw=$(arm-none-eabi-size $file)
+
+text=$(echo $raw | cut -d ' ' -f 7)
+data=$(echo $raw | cut -d ' ' -f 8)
+bss=$(echo $raw | cut -d ' ' -f 9)
+
+flash=$(($text + $data))
+ram=$(($data + $bss))
+
+print_region $flash $max_flash "Flash"
+print_region $ram $max_ram "RAM"
+```
+
+which gives us:
+```terminal
+francois-mba:with-libc francois$ ./get-fw-size build/with-libc.elf 0x40000
+0x8000
+Flash used: 10904 / 262144 (4%)
+RAM used: 8376 / 32768 (25%)
+```
+
+Stick that in your Makefiles, and you'll have the size as you build
+your firmware.
+
+### Digging into functions
+
+The above tells us *how much* code space we are using, but not *why*. To answer
+the latter, we need to go deeper.
+
+This is where another tool bundled with the GNU ARM toolchain comes in handy:
+`arm-none-eabi-nm`.
+
+`nm` allows us to print out all of the symbols in an ELF file, and with the
+`--size` option it will also print the size of each symbol, and even sort them
+by size if you specify `--size-sort`.
+
+So if we want to dig into what symbols consume the most code space, we simply
+do:
+
+```terminal
+francois-mba:with-libc francois$ arm-none-eabi-nm --print-size --size-sort --radix=d build/with-libc.elf | tail -30
```

Comment:
MAYBE: `-l` is nice to tack on too for line info

Reply:
Added a note, but didn't copy the output as it gets quite messy.